### ISSUE TITLE
False negative when quoted type appears in union syntax with UnionType operand

### ISSUE THREAD

**Issue by brianschubert** [microsoft/pyright#9500]
**Describe the bug**

Pyright does not reject quoted types in union syntax when combined with a `UnionType` operand.

**Code or Screenshots**

[[pyright playground]](https://pyright-play.net/?pyrightVersion=1.1.389&pythonVersion=3.12&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgIIA2SAhgM4CwAUIwB4CMAXJivgD5QBENGCH5Qx4iQGIoAUU4BVFEjAooNONyrMoAYyooUYfACNiUAK41iAEygB3JDAAWa4agxgExEPusBuCysoAEdzI2IaKCpwcxRbYm4kEDNiZgRkmhplVRYAJk5UXlcQKD5%2BYzgYCNEoaQAlWJgsFJBwEAAaKEMoBDgQJDQnfG82xkZ5TiIvCmpIgF4uIqEQFgBmBVKBCqqaGokpKAbEiBa2zu7e-sHh1twWABZOcsrqzfl9-frG5qgR3HOwD0%2BgMhr9biAgA)

```python
x1: int | "str"          # E: Union syntax cannot be used with string operand; use quotes around entire expression 
x2: int | str | "bytes"  # Runtime error, no pyright error

U: TypeAlias = int | str
x3: U | "bytes"          # Runtime error, no pyright error
x4: "bytes" | U          # Runtime error, no pyright error
```

**Command-line**
```shell
$ pyright example.py
example.py
 example.py:1:11 - error: Union syntax cannot be used with string operand; use quotes around entire expression (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 

$ pyright --version
pyright 1.1.389
```

**Comment by erictraut** [comment]
This isn't really a type error, but I think it's a reasonable enhancement request.

**Comment by loing-cdut** [comment]
Pyright is a static type checker developed by Microsoft for TypeScript and JavaScript, and it also supports type checking for Python code. In the information you provided, you pointed out an issue with Pyright when it comes to handling union types with quoted types. Specifically, it should not allow quoted string literals within union syntax.

In your examples, the definition of x1 is correct because it follows Pyright's rule that when using string literals in union types, the entire expression should be quoted. However, the definitions of x2, x3, and x4 will result in runtime errors because they attempt to use string literals in union types with type aliases or types without correctly quoting the entire expression.

Here is the English translation and explanation of your code:

python
# Pyright playground

# E: Union syntax cannot be used with string operand; use quotes around entire expression
x1: int | "str"  # Error: Union syntax cannot be used with string operand; use quotes around entire expression

# Runtime error, no pyright error
x2: int | str | "bytes"  

# Type alias for a union of int and str
U: TypeAlias = int | str

# Runtime error, no pyright error
x3: U | "bytes"  

# Runtime error, no pyright error
x4: "bytes" | U  
Command line output:

$ pyright example.py
example.py
 example.py:1:11 - error: Union syntax cannot be used with string operand; use quotes around entire expression (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 

$ pyright --version
pyright 1.1.389
In Pyright, the correct approach is to quote the entire union type, not individual string literals. For example:

python
# Correct usage
x5: int | str | "bytes"  # Correct: entire union is quoted
If you encounter such issues with Pyright and believe it is a bug or an area that needs improvement, you can submit an issue in Pyright's GitHub repository, providing your code examples and error messages so that developers can review and address the problem. Also, ensure you are using the latest version of Pyright, as this issue may have been fixed in newer releases.

**Comment by erictraut** [comment]
Looking at this more closely, there are a bunch of arcane (and undocumented) implementation details that dictate whether an error will be generated at runtime with the `|` operator. For example, `Union[str, int] | "bytes"` does not generate an error but `str | int | bytes` does. Replicating all of these runtime behaviors for all possible type forms without generating false positives in any cases would be challenging and error prone. It would also take a significant investment of time to get it right. 

For those reasons, I'm going to reject this enhancement request. As I said above, this isn't really a type checking feature anyway. I'm willing to reconsider this decision in the future if there's a signal from other pyright users (in the form of upvotes on this issue) that they'd like to see this.
